rfd/xxxx-headless-authentication.md

```diff
@@ -0,0 +1,386 @@
+---
+authors: Brian Joerger (bjoerger@goteleport.com)
+state: draft
+---
+
+# RFD TBD - Headless Authentication
+
+## Required Approvers
+
+* Engineering: @r0mant && @jakule
+* Security: @reed || @jentfoo
+* Product: @xinding33 || @klizhentas
+
+## What
+
+Support headless authentication for Teleport clients.
+
+Headless authentication is a non-interactive form of authentication that relies on a secondary party to authenticate a user. In our case, headless authentication will be used to authenticate a user on a remote machine using their credentials and MFA device on their local machine.
+
+## Why
+
+Some users have valid use cases for making `tsh` requests on a remote, sometimes shared, machine.
+
+For example:
+
+ 1. Connecting to Teleport Nodes.
+ 1. Performing `tsh scp` from one Teleport Node to another Teleport Node.
+
+However, `tsh` does not currently work well on remote machines for two reasons:
+
+ 1. Client credentials are usually saved to disk, where they may be overwritten, used, or stolen by anyone with access to the remote machine.
+ 1. Teleport widely uses MFA to authenticate client logins and actions, which may require direct access to a WebAuthn device. For example, with Per-session MFA, each SSH connection to a Teleport Node may require MFA verification.
+
+Headless authentication will provide a secure way to support these types of remote use cases.
+
+## Details
+
+### Security
+
+Headless authentication should fulfill the following design principles to ensure that the system can not be easily exploited by attackers, including attackers with root access to the remote machine.
+
+1. **The Client does not write any keys, certificates, or session data to disk on the remote machine**
+
+When using headless authentication, Teleport clients will generate a new private key in memory for the lifetime of the client. This private key will not be exported anywhere. Only its public key counterpart will be shared externally to get user certificates. Likewise, the user certificates will be held in memory for the lifetime of the client.
```

Comment:
This is neat, I wish there was a way to run tsh without ever writing credentials to disk, without necessarily reducing the scope of the credentials either.

Reply:
`tsh --add-keys-to-agent=only` is the closest we have today. But also, we could simply add a flag to use `MemClientStore`. I already made all (most?) of the changes necessary for this to work in https://github.com/gravitational/teleport/pull/19420.

Reply:
I'll take a look. 👍 

The original design for `tsh device enroll` said it was supposed to work like headless mode if the user wasn't already logged in, the idea being that during IT-driven manual enrollment we wouldn't leave powerful creds behind. I never did get to that, and also plans changed a bit, but it sounds like I could leverage what you've done to get that feature easily.

(Btw, `tsh device enroll` may be a good candidate for --headless.)